src/lib/sedlexing.ml

```diff
@@ -117,48 +140,84 @@ let refill lexbuf =
   end
   else lexbuf.len <- lexbuf.len + n
 
+let new_line lexbuf =
+  if lexbuf.curr_line != 0 then
+  lexbuf.curr_line <- lexbuf.curr_line + 1;
+  lexbuf.curr_bol <- lexbuf.pos + lexbuf.offset
+
 let next lexbuf =
   let i =
     if lexbuf.pos = lexbuf.len then
       if lexbuf.finished then eof
       else (refill lexbuf; lexbuf.buf.(lexbuf.pos))
     else lexbuf.buf.(lexbuf.pos)
   in
-  if i = eof then lexbuf.finished <- true else lexbuf.pos <- lexbuf.pos + 1;
+  if i = eof then lexbuf.finished <- true
+             else lexbuf.pos <- lexbuf.pos + 1;
+  (* '\n' = 10 *)
+  if i = 10 then new_line lexbuf;
   i
 
-let start lexbuf =
-  lexbuf.start <- lexbuf.pos;
-  lexbuf.marked_pos <- lexbuf.pos;
-  lexbuf.marked_val <- (-1)
 
 let mark lexbuf i =
   lexbuf.marked_pos <- lexbuf.pos;
+  lexbuf.marked_bol <- lexbuf.curr_bol;
+  lexbuf.marked_line <- lexbuf.curr_line;
   lexbuf.marked_val <- i
 
+let start lexbuf =
+  lexbuf.start_pos <- lexbuf.pos;
+  lexbuf.start_bol <- lexbuf.curr_bol;
+  lexbuf.start_line <- lexbuf.curr_line;
+  mark lexbuf (-1)
+
 let backtrack lexbuf =
   lexbuf.pos <- lexbuf.marked_pos;
+  lexbuf.curr_bol <- lexbuf.marked_bol;
+  lexbuf.curr_line <- lexbuf.marked_line;
   lexbuf.marked_val
 
 let rollback lexbuf =
-  lexbuf.pos <- lexbuf.start
+  lexbuf.pos <- lexbuf.start_pos;
+  lexbuf.curr_bol <- lexbuf.start_bol;
+  lexbuf.curr_line <- lexbuf.start_line
 
-let lexeme_start lexbuf = lexbuf.start + lexbuf.offset
+let lexeme_start lexbuf = lexbuf.start_pos + lexbuf.offset
 let lexeme_end lexbuf = lexbuf.pos + lexbuf.offset
 
-let loc lexbuf = (lexbuf.start + lexbuf.offset, lexbuf.pos + lexbuf.offset)
+let loc lexbuf = (lexbuf.start_pos + lexbuf.offset, lexbuf.pos + lexbuf.offset)
 
-let lexeme_length lexbuf = lexbuf.pos - lexbuf.start
+let lexeme_length lexbuf = lexbuf.pos - lexbuf.start_pos
 
 let sub_lexeme lexbuf pos len =
-  Array.sub lexbuf.buf (lexbuf.start + pos) len
+  Array.sub lexbuf.buf (lexbuf.start_pos + pos) len
 
 let lexeme lexbuf =
-  Array.sub lexbuf.buf (lexbuf.start) (lexbuf.pos - lexbuf.start)
+  Array.sub lexbuf.buf (lexbuf.start_pos) (lexbuf.pos - lexbuf.start_pos)
 
 let lexeme_char lexbuf pos =
-  lexbuf.buf.(lexbuf.start + pos)
-
+  lexbuf.buf.(lexbuf.start_pos + pos)
+
+let lexing_positions lexbuf =
+  let start_p = {
+    Lexing.pos_fname = lexbuf.filename;
+    pos_lnum = lexbuf.start_line;
+    pos_cnum = lexbuf.start_pos + lexbuf.offset;
+    pos_bol = lexbuf.start_bol;
+  } and curr_p = {
+    Lexing.pos_fname = lexbuf.filename;
+    pos_lnum = lexbuf.curr_line;
+    pos_cnum = lexbuf.pos + lexbuf.offset;
+    pos_bol = lexbuf.curr_bol;
+  } in
+  (start_p, curr_p)
+
+let convert_for_menhir lexer' lexbuf =
+  let lexer () =
+    let token = lexer' lexbuf in
+    let (start_p, curr_p) = lexing_positions lexbuf in
+    (token, start_p, curr_p)
+  in lexer
```

Comment:
What happens at the end of stream ? Does the menhir API specifies something ?

Reply:
In our package, eof has length 0 at file end, which is consist of ocamllex.
menhir API says only that position information is compatible with ocamlyacc and I could not found document of ocamlyacc about this.